lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/TaskManager.cs

```diff
@@ -0,0 +1,474 @@
+            }
+        }
+
+        /// <summary>
+        /// Returns the number of tasks in the task collection
+        /// </summary>
+        internal int NumberOfTasks
+        {
+            get { return _tasks.Count; }
+        }
+
+        /// <summary>
+        /// This method is called when receiving IRunningTask event during system SubmittingTasks state.
+        /// Adds the IRunningTask to the running tasks collection and update the task state to TaskRunning.
+        /// Throws IMRUSystemException if running tasks already contains this task or tasks collection doesn't contain this task.
+        /// </summary>
+        /// <param name="runningTask"></param>
+        internal void RecordRunningTask(IRunningTask runningTask)
+        {
+            if (_runningTasks.ContainsKey(runningTask.Id))
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture, "The task [{0}] already running.", runningTask.Id);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            if (!_tasks.ContainsKey(runningTask.Id))
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture, "The task [{0}] doesn't exist.", runningTask.Id);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            _runningTasks.Add(runningTask.Id, runningTask);
+            UpdateState(runningTask.Id, TaskStateEvent.RunningTask);
+        }
+
+        /// <summary>
+        /// This method is called at the beginning of the recovery.
+        /// Clears the task collection, running task collection and resets the number of application error. 
+        /// </summary>
+        internal void Reset()
+        {
+            _tasks.Clear();
+            _runningTasks.Clear();
+            _numberOfAppError = 0;
+        }
+
+        /// <summary>
+        /// This method is called when receiving ICompletedTask event during task running or system shutting down.
+        /// Removes the task from running tasks
+        /// Changes the task state from RunningTask to CompletedTask
+        /// </summary>
+        /// <param name="taskId"></param>
+        internal void RecordCompletedTask(string taskId)
+        {
+            RemoveRunningTask(taskId);
+            UpdateState(taskId, TaskStateEvent.CompletedTask);
+        }
+
+        /// <summary>
+        /// This method is called when receiving IFailedTask event during task submitting or running
+        /// Removes the task from running tasks if the task was running
+        /// Updates the task state to fail based on the error message in the failed task
+        /// </summary>
+        /// <param name="failedTask"></param>
+        internal void RecordFailedTaskDuringRunningOrSubmissionState(IFailedTask failedTask)
+        {
+            //// Remove the task from running tasks if it exists there
+            _runningTasks.Remove(failedTask.Id);
```

Comment:
This method can be called in submitting state when the task is not in _trunningTasks yet. _runningTasks.Remove will not remove the task if it doesn't exist and returns false. Here I just simply ignore the return value so that it works for both running and submitting state. 
In other cases, I do want to check if  _trunningTasks  contains the task id. If not throw exception.